### DISCUSSION TITLE
Partial Monte-Carlo dropout

### DISCUSSION THREAD

**Discussion by pieterblok** [baal-org/baal#252] [Ideas]
Monte-Carlo dropout is a simple but time-consuming procedure when determining the uncertainty score of an image. To alleviate this problem, I have implemented the so-called Partial Monte-Carlo dropout (PMCD) procedure in Detectron2 Active Learning (see line 173 in https://github.com/pieterblok/maskal/blob/maskAL/active_learning/sampling/montecarlo_dropout.py). 

With PMCD an image is inferred only once through the neural network parts that don't have dropout. Then, the resulting features from these parts are copy-paste to the network parts with dropout where the actual repeated inference happens. PMCD is highly scalable and significantly faster than inferring the same image multiple times through the entire network. I was wondering if I can implement this PMCD with one of your developers in BAAL. @Dref360 @parmidaatg

**Comment by parmidaatg** [comment]
Hi @pieterblok, thank you so much for your comment and yes definitely please go ahead and make a PR. If you were interested you can then maybe generate some metrics to compare the two using BaaL and we post it in our tutorial. Otherwise I can do that part :)

**Comment by Dref360** [comment]
cc @orangetoaster 

That sounds great!

**Comment by pieterblok** [comment]
@parmidaatg @Dref360 @orangetoaster 

Perfect! I will have a deep dive into the BAAL source code to look how I can integrate such method. If I'm stuck somewhere or lost in integration, I will ask in this Discussion page.

**Comment by Dref360** [comment]
@pieterblok and I met on Friday to discuss this issue.
He proposed a great idea where users would implement a custom function that would have the following signature:

```python
def montecarlo_forward(x : Tensor[BatchSize, ...], iterations:int) -> Tensor[BatchSize, ..., ITERATION]:
pass
```

This would make the case where we have "replicate_in_memory=False" much faster. Pieter shared a speedup of 20 to 40% on his models.



We also discussed another way of doing it with a LRU cache. 

This PR is the first implementation of such proposal: https://github.com/baal-org/baal/pull/254

Let me know what you think!

**Reply by pieterblok** [reply]
> @pieterblok and I met on Friday to discuss this issue. He proposed a great idea where users would implement a custom function that would have the following signature:
> 
> ```python
> def montecarlo_forward(x : Tensor[BatchSize, ...], iterations:int) -> Tensor[BatchSize, ..., ITERATION]:
> pass
> ```
> 
> This would make the case where we have "replicate_in_memory=False" much faster. Pieter shared a speedup of 20 to 40% on his models.
> 
> We also discussed another way of doing it with a LRU cache.
> 
> This PR is the first implementation of such proposal: #254
> 
> Let me know what you think!

@Dref360 thank you very much for your time and input!

Indeed, my suggestion is as follows. Assuming this standard forward pass:
```python
def forward(self, x):
        
    skip_connections = []
    size_in = x.size()
    
    
    for i in range(len(self.base)):
        x = self.base[i](x)
        if i in self.shortcut_layers:
            skip_connections.append(x)
    out = x
    
    for i in range(self.n_blocks):
        skip = skip_connections.pop()
        out = self.transUpBlocks[i](out, skip, True)
        out = self.conv1x1_up[i](out)
        out = self.denseBlocksUp[i](out)
    
    out = self.finalConv(out)
    
    out = F.interpolate(
                        out,
                        size=(size_in[2], size_in[3]),
                        mode="bilinear",
                        align_corners=True)
    return out
```

Then copy-paste this forward pass to a new function and name that function **montecarlo_forward** (as part of your neural network class). Make it iterable for only the network parts that have dropout (in this example it's the last layer **out = self.finalConv(out)**):
```python
def montecarlo_forward(self, x, iterations):

    skip_connections = []
    size_in = x.size()
    
    
    for i in range(len(self.base)):
        x = self.base[i](x)
        if i in self.shortcut_layers:
            skip_connections.append(x)
    out = x
    
    for i in range(self.n_blocks):
        skip = skip_connections.pop()
        out = self.transUpBlocks[i](out, skip, True)
        out = self.conv1x1_up[i](out)
        out = self.denseBlocksUp[i](out)
    
    out_list = []
    for _ in range(iterations):
        out_finalconv = self.finalConv(out)
        
        out_final = F.interpolate(
                            out_finalconv,
                            size=(size_in[2], size_in[3]),
                            mode="bilinear",
                            align_corners=True)

        out_list.append(out_final)

    return out_list
```

Change this code
https://github.com/baal-org/baal/blob/cea4693ff600bc53885d8f5b44f6d0991bc74d15/baal/modelwrapper.py#L390-L392

to:
```python
else:
    try:
        func_list = dir(self.model.module)
    except:
        func_list = []
        
    if "montecarlo_forward" in func_list:
        out = self.model.module.montecarlo_forward(data, iterations)
    else:
        out = [self.model(data) for _ in range(iterations)]
    out = _stack_preds(out)
```

That should make the repeated image analysis during Monte Carlo sampling faster (between 25% and 60% for this network).

**Reply by Dref360** [reply]
Thank you for your PR Pieter.
So we now have two potential approaches: #255 and #254.

I'll write down the pros and cons of each and we can then take a decision.

#255 Support through an interface
Pros:
* The user has all the control
* No memory overhead
 
Cons:
* The user has to implement the interface, but our target userbase is technical so that shouldn't be an issue.

#254 Use an LRU cache
Pros:
* Plug and play

Cons:
* Has a slight memory overhead, but not worse than the backprop overhead, so shouldn't be an issue for most users.

I think #254 is simpler to use so I would prefer it, but I would love to gather more impressions before we make a decision.

**Reply by pieterblok** [reply]
> Thank you for your PR Pieter.
> 
> So we now have two potential approaches: #255 and #254.
> 
> 
> 
> I'll write down the pros and cons of each and we can then take a decision.
> 
> 
> 
> #255 Support through an interface
> 
> Pros:
> 
> * The user has all the control
> 
> * No memory overhead
> 
>  
> 
> Cons:
> 
> * The user has to implement the interface, but our target userbase is technical so that shouldn't be an issue.
> 
> 
> 
> #254 Use an LRU cache
> 
> Pros:
> 
> * Plug and play
> 
> 
> 
> Cons:
> 
> * Has a slight memory overhead, but not worse than the backprop overhead, so shouldn't be an issue for most users.
> 
> 
> 
> I think #254 is simpler to use so I would prefer it, but I would love to gather more impressions before we make a decision.
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 

@Dref360 thanks for the update. Your suggestion, #254, is probably preferable, however I was not able to test it yet. I hope to do it sometime this week. Maybe I will send you a private message or email to get it up and running for the semantic segmentation. Thanks again!

**Reply by pieterblok** [reply]
@Dref360 I have conducted some first tests on my laptop computer (Lenovo P53, with NVIDIA Quadro RTX 5000 GPU)

On my laptop, the vgg example (#254) is just a fraction faster (probably not significant). I have implemented the code below based on #254, however I'm not sure whether the regular implementation is the correct one (this was missing in #254):

```python
import torch
from torchvision.models import vgg16
from tqdm import tqdm

from baal.bayesian.caching_utils import MCCachingModule
from baal.bayesian.dropout import MCDropoutModule

vgg = vgg16().cuda()
vgg.eval()

input = torch.randn(10, 3, 224, 224).cuda()

# Cached : 1:08
with MCCachingModule(vgg) as model:
    with MCDropoutModule(model) as model_2:
        [model_2(input).detach().cpu() for _ in tqdm(range(1000))]

# Regular: 1:10
with MCDropoutModule(vgg) as model_3:
    [model_3(input).detach().cpu() for _ in tqdm(range(1000))]  
``` 

Furthermore, I have tested the MCCachingModule in our semantic segmentation active learning framework. In that code, I have implemented the MCCachingModule before the ModelWrapper (we don't have a MCDropoutModule, but a custom builder called ConvLayerDO that rebuilds the last convolutional layer by adding a nn.Dropout2D before it). We are using [FCHardNet](https://github.com/PingoLH/FCHarDNet/blob/master/ptsemseg/models/hardnet.py). Some snippets of our code:
```python
self.model = hardnet
self.model.finalConv = ConvLayerDO(self.model.finalConv, probability=dropout_probability)
self.model = patch_module(self.model)

self.model = MCCachingModule(self.model)
self.model = ModelWrapper(self.model, loss_fn, replicate_in_memory=False)
```

_Sampling times on the first 30 images of our pool:_
| Sampling method   |     Number of sampled images     |  Time (min:sec) |
|:----------:|:-------------:|:------:|
| Default Monte-Carlo method |  30 | 1:37 |
| Partial Monte-Carlo method (#255) |    30  |   1:13 |
| MCCachingModule (#254) | 30 |  2:01 |

I don't understand this. Am I doing something wrong?

**Reply by Dref360** [reply]
How interesting! Thank you for testing it on your side! Will dig deeper.

**Reply by Dref360** [reply]
Hello! I'm finally back from Italy and I'm determined to figure this out.

Could you tell me what is your Pytorch/CUDA version? I suspect the `torch.equal` has been optimized recently. I personally have CUDA 11.7 and pytorch 2.0. I have a GTX 1060 on my laptop.


Also, I have a new test that is more "real-life like".

```python
from torchvision.datasets import CIFAR10
from torchvision.models import vgg16
from torchvision.transforms import ToTensor

from baal.bayesian.caching_utils import MCCachingModule
from baal.bayesian.dropout import MCDropoutModule
from baal.modelwrapper import ModelWrapper

vgg = vgg16().cuda()
vgg.eval()

ds = CIFAR10('/tmp', train=False, transform=ToTensor(), download=True)

# Regular: 2:58
# Cached : 50 seconds
with MCCachingModule(vgg) as model:
    with MCDropoutModule(model) as model_2:
        wrapper = ModelWrapper(model_2, None, replicate_in_memory=False)
        wrapper.predict_on_dataset(ds, batch_size=32, iterations=20, use_cuda=True)

with MCDropoutModule(vgg) as model_2:
    wrapper = ModelWrapper(model_2, None, replicate_in_memory=False)
    wrapper.predict_on_dataset(ds, batch_size=32, iterations=20, use_cuda=True)

```

**Reply by pieterblok** [reply]
> Hello! I'm finally back from Italy and I'm determined to figure this out.
> 
> Could you tell me what is your Pytorch/CUDA version? I suspect the `torch.equal` has been optimized recently. I personally have CUDA 11.7 and pytorch 2.0. I have a GTX 1060 on my laptop.
> 
> Also, I have a new test that is more "real-life like".
> 
> ```python
> from torchvision.datasets import CIFAR10
> from torchvision.models import vgg16
> from torchvision.transforms import ToTensor
> 
> from baal.bayesian.caching_utils import MCCachingModule
> from baal.bayesian.dropout import MCDropoutModule
> from baal.modelwrapper import ModelWrapper
> 
> vgg = vgg16().cuda()
> vgg.eval()
> 
> ds = CIFAR10('/tmp', train=False, transform=ToTensor(), download=True)
> 
> # Regular: 2:58
> # Cached : 50 seconds
> with MCCachingModule(vgg) as model:
>     with MCDropoutModule(model) as model_2:
>         wrapper = ModelWrapper(model_2, None, replicate_in_memory=False)
>         wrapper.predict_on_dataset(ds, batch_size=32, iterations=20, use_cuda=True)
> 
> with MCDropoutModule(vgg) as model_2:
>     wrapper = ModelWrapper(model_2, None, replicate_in_memory=False)
>     wrapper.predict_on_dataset(ds, batch_size=32, iterations=20, use_cuda=True)
> ```

@Dref360 welcome back, and thanks for the update! 

I'm using Pytorch version 1.8.0 with CUDA 10.1. Compared to your versions pretty outdated. I will try to get my laptop or another computer updated, so that I can replicate your experiment. I will immediately try our active learning with this cached approach, that might take some more time.

**Reply by pieterblok** [reply]
@Dref360 fortunately my colleague had already a system with Pytorch 2.0.1 and CUDA 11.7 installed, so it was easy to replicate this test. We have found the following:

"The MCCachingModule is indeed faster, approximately twice as fast. The total it/s of MCCachingModule is 20.72 while the one without caching was 10.83."

Next step is testing the prediction speed in our specific active learning application. 

So far, so good... happy that we were able to reproduce your code with the newest libraries!

**Reply by Dref360** [reply]
@pieterblok I was able to run some tests on a UNET. The results are less impressive. I'm able to improve the results by 45% using the following script.


```python
from typing import List

import segmentation_models_pytorch as smp
import torch
from segmentation_models_pytorch.base import SegmentationHead
from tqdm import tqdm

from baal.bayesian.caching_utils import MCCachingModule
from baal.bayesian.dropout import MCDropoutModule


DS_SIZE=100
ITERATIONS=20

def add_dropout(
        model: smp.Unet,
        decoder_channels: List[int] = (256, 128, 64, 32, 16),
        classes=1,
        activation=None,
):
    seg_head = SegmentationHead(
        in_channels=decoder_channels[-1],
        out_channels=classes,
        activation=activation,
        kernel_size=3,
    )
    model.add_module("segmentation_head", seg_head)
    model.initialize()


unet = smp.Unet(
    encoder_name="resnext50_32x4d",
    encoder_depth=5,
    encoder_weights="imagenet",
    decoder_use_batchnorm=False,
    classes=10,
)
# Add a Dropout layerto use MC-Dropout
add_dropout(unet, classes=10, activation=None)
unet = unet.eval().cuda()

# Regular: 1:30
# Cached: 52 seconds
with MCCachingModule(unet) as model:
    with MCDropoutModule(model) as model_2:
        for batch in tqdm(range(DS_SIZE)):
            input = torch.randn(10, 3, 224, 224).cuda()
            [model_2(input).detach().cpu() for _ in range(ITERATIONS)]

```

My hypothesis is that the new MCCachingModule doesn't cache concatenation hence the small speedup. This is something we could try to improve in the future. 50% is still a nice speedup. 

I would propose merging MCCachingModule

**Reply by pieterblok** [reply]
@Dref360 perfect! Indeed that makes sense. Thanks for the effort.
